File: src/Library/ObjectModel/WirelessNetwork.vb
Public Class WirelessNetwork
   Inherits Pixa.Soundbridge.Library.SoundbridgeListObject

    Private _connected As Boolean
    Private _selected As Boolean

    Public Event SelectedChanged As EventHandler

#Region " Classes "
    Public Class SignalQuality

        Private _time As DateTime
        Private _quality As Integer
        Private _signal As Integer
        Private _noise As Integer

        Public Sub New(ByVal quality As Integer, ByVal signal As Integer, ByVal noise As Integer)
            _time = DateTime.Now
            _quality = quality
            _signal = signal
            _noise = noise
        End Sub

        Public ReadOnly Property Noise() As Integer
            Get
                Return _noise
            End Get
        End Property

        Public ReadOnly Property Quality() As Integer
            Get
                Return _quality
            End Get
        End Property

        Public ReadOnly Property Signal() As Integer
            Get
                Return _signal
            End Get
        End Property

        Public ReadOnly Property Time() As DateTime
            Get
                Return _time
            End Get
        End Property
    End Class
#End Region

    Friend Sub New(ByVal client As ISoundbridgeClient, ByVal index As Integer, ByVal name As String, ByVal connected As Boolean, ByVal selected As Boolean)
      MyBase.New(client, index, name)
    End Sub

    Public ReadOnly Property Connected() As Boolean
        Get
            Return _connected
        End Get
    End Property

    Public Property Selected() As Boolean
        Get
            Return _selected
        End Get
        Friend Set(ByVal value As Boolean)
            _selected = value
        End Set
    End Property

    Public Function GetSignalQuality() As SignalQuality
        If Not Connected Then Throw New InvalidOperationException("Can't get the signal quality when the soundbridge isn't connected to the network")

        Dim quality As Integer
        Dim signal As Integer
        Dim noise As Integer

        For Each s As String In Client.GetWiFiSignalQuality
            If s = "OK" Then Continue For

            Dim iNumberLeft As Integer = s.Length - 1

            While Char.IsNumber(s, iNumberLeft) Or s(iNumberLeft) = "-"
                iNumberLeft -= 1
            End While

            Dim sNumber As Integer = s.Substring(iNumberLeft + 1)
            Dim i As Integer

            If Integer.TryParse(sNumber, i) Then
                If s.StartsWith("quality") Then
                    quality = i
                ElseIf s.StartsWith("signal") Then
                    signal = i
                ElseIf s.StartsWith("noise") Then
                    noise = i
                End If
            End If
        Next

        Return New SignalQuality(quality, signal, noise)
    End Function

    Public Sub [Select](ByVal password As String)
        Dim r As String

        r = Client.SetWiFiNetworkSelection(Index)
        If r <> "OK" Then ExceptionHelper.ThrowCommandReturnError("SetWiFiNetworkSelection", r)

        If password <> "" Then
            r = Client.SetWiFiPassword(password)
            If r <> "OK" Then ExceptionHelper.ThrowCommandReturnError("SetWiFiPassword", r)
        End If
    End Sub
End Class

File: src/Pixa.Soundbridge/WirelessNetwork.cs
using Pixa.Soundbridge.Client;
using System;

namespace Pixa.Soundbridge
{
    public class WirelessNetwork : SoundbridgeListObject
    {
        private bool _connected;
        private bool _selected;

        public event EventHandler SelectedChanged;

        #region  Classes 
        public class SignalQuality
        {
            private DateTime _time;
            private int _quality;
            private int _signal;
            private int _noise;

            public SignalQuality(int quality, int signal, int noise)
            {
                _time = DateTime.Now;
                _quality = quality;
                _signal = signal;
                _noise = noise;
            }

            public int Noise
            {
                get
                {
                    return _noise;
                }
            }

            public int Quality
            {
                get
                {
                    return _quality;
                }
            }

            public int Signal
            {
                get
                {
                    return _signal;
                }
            }

            public DateTime Time
            {
                get
                {
                    return _time;
                }
            }
        }
        #endregion

        internal WirelessNetwork(ISoundbridgeClient client, int index, string name, bool connected, bool selected) : base((Soundbridge)client, index, name)
        {
        }

        public bool Connected
        {
            get
            {
                return _connected;
            }
        }

        public bool Selected
        {
            get
            {
                return _selected;
            }

            internal set
            {
                _selected = value;
            }
        }

        public SignalQuality GetSignalQuality()
        {
            if (!Connected)
                throw new InvalidOperationException("Can't get the signal quality when the soundbridge isn't connected to the network");
            var quality = default(int);
            var signal = default(int);
            var noise = default(int);
            foreach (string s in Client.GetWiFiSignalQuality())
            {
                if (s == "OK")
                    continue;
                int iNumberLeft = s.Length - 1;
                while (char.IsNumber(s, iNumberLeft) | s[iNumberLeft] == '-')
                    iNumberLeft -= 1;
                int sNumber = int.Parse(s.Substring(iNumberLeft + 1));
                int i;
                if (int.TryParse(sNumber.ToString(), out i))
                {
                    if (s.StartsWith("quality"))
                    {
                        quality = i;
                    }
                    else if (s.StartsWith("signal"))
                    {
                        signal = i;
                    }
                    else if (s.StartsWith("noise"))
                    {
                        noise = i;
                    }
                }
            }

            return new SignalQuality(quality, signal, noise);
        }

        public void Select(string password)
        {
            string r;
            r = Client.SetWiFiNetworkSelection(Index);
            if (r != "OK")
                ExceptionHelper.ThrowCommandReturnError("SetWiFiNetworkSelection", r);
            if (!string.IsNullOrEmpty(password))
            {
                r = Client.SetWiFiPassword(password);
                if (r != "OK")
                    ExceptionHelper.ThrowCommandReturnError("SetWiFiPassword", r);
            }
        }
    }
}